h5py/_npystrings.pyx

```diff
@@ -0,0 +1,254 @@
+    for a HDF5 variable-width string dataset, into a NpyString array
+    (NumPy's native variable-width strings dtype).
+
+    Parameters
+    ----------
+    space : hid_t
+        HDF5 dataspace ID for the output non-contiguous buffer
+        containing packed NpyStrings
+    contig : void * (actually a char **)
+        Input buffer for the contiguous zero-terminated char **
+    noncontig : void *
+        Output buffer for the non-contiguous packed NpyStrings
+    descr : PyArray_Descr * (actually a PyArray_StringDTypeObject *)
+        NumPy dtype descriptor for the NpyStrings
+    npoints : size_t
+        Number of points to copy
+
+    Notes
+    -----
+    This function deep-copies the input zero-terminated strings.
+    Memory management for the outputs is handled by NumPy.
+    """
+    cdef vstrings_scatter_t info
+    info.i = 0
+    info.contig = <const char**>contig
+    info.allocator = NpyString_acquire_allocator(<PyArray_StringDTypeObject *>descr)
+    if info.allocator is NULL:
+        raise RuntimeError("Failed to acquire string allocator")
+
+    # Disregard mtype from _proxy::dset_rw.
+    # The memory type is actually npy_packed_static_string
+    # (an opaque struct *typically* 16 bytes per point),
+    # and not HDF5 variable length strings (char*; 8 bytes per point).
+    tid = H5Tcreate(H5T_OPAQUE, SIZEOF_NPY_PACKED_STATIC_STRING)
+
+    # Read char*[] (zero-terminated) from h5py
+    # and deep-copy to npy_packed_static_string[] for numpy
+    H5Diterate(noncontig, tid, space, vstrings_scatter_cb, &info)
+    NpyString_release_allocator(info.allocator)
+    H5Tclose(tid)
+
+
+cdef herr_t vstrings_gather_cb(
+    void *elem, hid_t type_id, unsigned ndim,
+    const hsize_t *point, void *operator_data
+) except -1:
+    cdef vstrings_gather_t *info = <vstrings_gather_t *>operator_data
+    cdef npy_static_string *unpacked = info[0].unpacked + info[0].i
+    # Obtain a reference to the string (NOT zero-terminated) and its size
+    res = NpyString_load(info[0].allocator, <npy_packed_static_string*>elem, unpacked)
+    info[0].i += 1
+    return res
+
+def vstrings_gather(hid_t space, size_t contig, size_t noncontig, size_t descr,
+                    size_t npoints):
+    """Pure-python API interface to _proxy.pyx. This is necessary to
+    allow this module to be conditionally imported, allowing numpy 1.x to
+    continue working everywhere else.
+    """
+    return <size_t>_vstrings_gather(space, <void *>contig, <void *>noncontig,
+                                    <PyArray_Descr *>descr, npoints)
+
+
+cdef char * _vstrings_gather(hid_t space, void *contig, void *noncontig,
+                             PyArray_Descr *descr, size_t npoints):
+    """Convert a NpyString array (NumPy's native variable-width strings dtype)
+    to a zero-terminated char**, which is the in-memory representation for a
+    HDF5 variable-width string dataset.
```

Comment:
It occurs to me that this is the same operation as asking NumPy to do a cast to a fixed-width byte array (which outputs UTF-8 data).

I wonder if that would be more maintainable and/or faster than trying to do the copies yourself.

Reply:
I don't understand. Isn't a fixed-witdth array padded with NULLs? That would potentially cause a very high peak in memory usage?